packages/framework/client-logger/external-telemetry/src/container/telemetryProducer.ts

```diff
@@ -0,0 +1,107 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	IContainer,
+	IContainerEvents,
+	type ICriticalContainerError,
+} from "@fluidframework/container-definitions";
+import {
+	ContainerConnectedTelemetry,
+	ContainerTelemetryEventNames,
+	IContainerTelemetry,
+	ContainerDisconnectedTelemetry,
+	ContainerClosedTelemetry,
+	ContainerAttachingTelemetry,
+	ContainerAttachedTelemetry,
+	type ContainerTelemetryEventName,
+} from "./containerTelemetry";
+import { ContainerSystemEventName, ContainerSystemEventNames } from "./containerSystemEvents";
+
+/**
+ * This class produces {@link IContainerTelemetry} from raw container system events {@link IContainerEvents}.
+ * The class contains different helper methods for simplifying and standardizing logic for adding additional information necessary
+ * to produce different {@link IContainerTelemetry}.
+ *
+ * @internal
+ */
+export class ContainerEventTelemetryProducer {
+	constructor(private container: IContainer) {}
+
+	public produceTelemetry(
+		eventName: ContainerSystemEventName,
+		payload?: any,
+	): IContainerTelemetry | undefined {
+		let telemetry: IContainerTelemetry | undefined = undefined;
+		switch (eventName) {
+			case ContainerSystemEventNames.CONNECTED:
+				telemetry = this.produceConnectedTelemetry(payload);
+				return telemetry;
+			case ContainerSystemEventNames.DISCONNECTED:
+				telemetry = <ContainerDisconnectedTelemetry>(
+					this.produceBasicContainerTelemetry(ContainerTelemetryEventNames.DISCONNECTED)
+				);
+				break;
+			case ContainerSystemEventNames.CLOSED:
+				telemetry = this.produceClosedTelemetry(payload);
+				break;
+			case ContainerSystemEventNames.ATTACHED:
+				telemetry = <ContainerAttachedTelemetry>(
+					this.produceBasicContainerTelemetry(ContainerTelemetryEventNames.ATTACHED)
+				);
+				break;
+			case ContainerSystemEventNames.ATTACHING:
+				telemetry = <ContainerAttachingTelemetry>(
+					this.produceBasicContainerTelemetry(ContainerTelemetryEventNames.ATTACHING)
+				);
+				break;
+			default:
+				break;
+		}
+		return telemetry;
+	}
+
+	private produceBasicContainerTelemetry = <T extends IContainerTelemetry>(
+		eventName: ContainerTelemetryEventName,
+	): T => {
+		return {
+			eventName,
+			containerId: this.getContainerId(),
+			documentId: this.getDocumentId(),
+		} as T;
+	};
+
+	private produceConnectedTelemetry = (payload?: {
+		clientId: string;
+	}): ContainerConnectedTelemetry => {
+		return {
+			eventName: ContainerTelemetryEventNames.CONNECTED,
+			containerId: payload?.clientId ?? this.getContainerId(),
```

Comment:
Related to previous comment, let's not mix these two things. We had discussed that `containerId` could be a UUID generated by this layer as long as it is consistent for the same container.

Also, have we checked if clientId is present in `container` at this time? If it is, I think I'd rather just let the connected telemetry go through the same path as the other basic events.

Reply:
Do we want to call this 'clientId' on the telemetry itself then?

> Also, have we checked if clientId is present in container at this time? If it is, I think I'd rather just let the connected telemetry go through the same path as the other basic events.

I have not confirmed this yet, just trying to be safe.

Reply:
> Do we want to call this 'clientId' on the telemetry itself then?

Yeah but I think both are valuable. A container might disconnect and reconnect while it is alive, which will result in the clientId changing, but it's still the same container and it's useful to know that, so having a persisent `containerId` makes sense. I'd say we can assign a UUID to a container when we "receive it", and make that the `containerId`. There were discussions about how to correlate that with the `containerId` we currently have in the internal telemetry, which isn't exposed anywhere, so that's something we'll want to look into soon, but for this iteration I think it's fine if they don't match.